python/cog/server/worker.py

```diff
@@ -328,28 +342,57 @@ def _setup(self, redirector: StreamRedirector) -> None:
                     raise
                 self._events.send(done)
 
-    def _loop(self, redirector: StreamRedirector) -> None:
+    def _loop(
+        self,
+        predict: Callable[..., Any],
+        redirector: StreamRedirector,
+    ) -> None:
         with redirector:
             while True:
                 ev = self._events.recv()
-                if isinstance(ev, Shutdown):
+                if isinstance(ev, Cancel):
+                    continue  # Ignored in sync predictors.
+                elif isinstance(ev, Shutdown):
                     break
-                if isinstance(ev, PredictionInput):
-                    self._predict(ev.payload, redirector)
+                elif isinstance(ev, PredictionInput):
+                    self._predict(ev.payload, predict, redirector)
                 else:
                     print(f"Got unexpected event: {ev}", file=sys.stderr)
 
+    async def _aloop(
+        self,
+        predict: Callable[..., Any],
+        redirector: AsyncStreamRedirector,
+    ) -> None:
+        # Unwrap and replace the events connection with an async one.
+        assert isinstance(self._events, LockedConnection)
+        self._events = AsyncConnection(self._events.connection)
```

Comment:
it may be cleaner to have this be a separate variable name instead of making _events.recv either sync or async